js/service/attachmentservice.js

```diff
@@ -66,4 +71,81 @@ define(function(require) {
 		return Promise.resolve($.ajax(url));
 	}
 
+	/**
+	 * @param {File} file
+	 * @param {function} localAttachmentModel
+	 * @returns {Promise}
+	 */
+	function uploadLocalAttachment(file, localAttachmentModel) {
+		var defer = $.Deferred();
+		var fd = new FormData();
+		fd.append('attachment', file);
+
+		var progressCallback = localAttachmentModel.onProgress;
+
+		var url = OC.generateUrl('/apps/mail/attachments');
+		$.ajax({
+			url: url,
+			type: 'POST',
+			xhr: function() {
+				var customXhr = $.ajaxSettings.xhr();
+				// save the xhr into the model in order to :
+				//  - distinguish upload and nextcloud file attachments
+				//  - keep the upload status for later use
+				localAttachmentModel.set('uploadRequest', customXhr);
+				// and start the request
+				if (customXhr.upload && _.isFunction(progressCallback)) {
+					customXhr.upload.addEventListener(
+						'progress',
+						progressCallback.bind(localAttachmentModel),
+						false);
+				}
+				return customXhr;
+			},
+			data: fd,
+			processData: false,
+			contentType: false,
+		}).done(function(data) {
+			defer.resolve(localAttachmentModel, data.id);
+		}).fail(function(err) {
+			defer.reject(localAttachmentModel);
+		});
+
+		return defer.promise();
```

Comment:
in meantime we've moved to native ES6 promises. Those jQuery deferreds should be replaced hence :)